app/models/category.rb

```diff
@@ -15,4 +15,14 @@ class Category < ApplicationRecord
   validates :amount, :name, presence: true
   validates :amount, numericality: {greater_than_or_equal_to: 0}
   validates :name, length: {minimum: 2, maximum: 40}
+
+  def descendents
+    sub_categories.map do |sub_category|
+      [sub_category] + sub_category.sub_categories + sub_category.descendents
+    end.flatten.uniq
+  end
+
+  def self_and_descendents
+    [self] + descendents
+  end
```

Comment:
I use this methods in a decorator, so it looks for me better to have this in a model to have a straight connection with record. For me it will be overcomplicated to use this with service.

Reply:
@sinpai You used this methods to find all subcategories, and in decorators used amount of these records. It good logic for one service. If need, I can help you make it happen in the evening or tomorrow.